crates/red_knot_python_semantic/src/types.rs

```diff
@@ -220,6 +220,9 @@ fn declarations_ty<'db>(
     } else {
         first
     };
+    // Make sure not to emit spurious errors relating to `Type::Todo`,
+    // since we only infer this type due to a limitation in our current model
```

Comment:
Hmm, from just reading this comment, it isn't clear to me why we are only filtering `Todo` types? Why aren't we filtering out `Unknown` types? I thought the two are equivalent except that they indicate a different source.

Reply:
That's a fair point. I think we should never infer a variable that has a declaration as being of type `Unknown` unless they had an invalid type annotation of some kind, in which case we should have already emitted a diagnostic. So probably we should also filter out `Unknown` types here, to avoid cascading errors from one mistake.

> Would it be possible to remove the `retain` call and instead avoid adding the types to `conflicting` on line 214

makes sense

Reply:
Ah, no. We do need to keep `Unknown` in `conflicting`, or we won't complain about this case:

```py
if flag:
    x: int
x = 1  # error: [conflicting-declarations] "Conflicting declared types for `x`: Unknown, int"
```

Currently every undeclared symbol in our model is marked as being "implicitly declared as `Unknown`"; this is how we detect that a symbol is only annotated in one branch.

I also don't think it's _so_ awful if a user gets two errors rather than one when they've got an invalid annotation in their code